add reservoir timers

maciuszek opened the pull request:
Add a reservoir implementation for timers. This relies on a new timer type, and using this type should allow for a more controlled stat load from timers.

maciuszek committed:
prototype a simple reservoir for timers

maciuszek commented on stats_test.go:
```diff
@@ -89,10 +90,11 @@ func TestValidateTags(t *testing.T) {
-	counter = sink.record
-	if !strings.Contains(counter, expected) {
-		t.Error("wanted counter value of test.___f=i:1|c, got", counter)
+	expected = "test.__host=i:1|c"
```
Note: this test change is out of scope of this work but it was previously volatile with the order of reserved_tag vs test.__host not being deterministic

maciuszek committed:
clarify comment

bshramin committed:
Move s.timerCount inside timer.Range

maciuszek committed:
fix test messages

maciuszek committed:
fix spelling

maciuszek committed:
Merge remote-tracking branch 'origin/master' into mattkuzminski/add-timer-reservoir

maciuszek committed:
re-design with timer reservoirs correctly independent per mertic
re-design with the notion of sample rate appended to the metric

maciuszek committed:
add some more todos

maciuszek committed:
clean up redundant code
improve naming

maciuszek committed:
some more clean up

maciuszek commented on sink.go:
```diff
@@ -6,6 +6,7 @@ type Sink interface {
 	FlushCounter(name string, value uint64)
 	FlushGauge(name string, value uint64)
 	FlushTimer(name string, value float64)
+	FlushAggregatedTimer(name string, value, sampleRate float64)
```
Because of this we'll need a new major release. 

The library usage should be backwards compatible, and additionally there is a feature flag to control the new behaviour, however if anything is implementing this interface it'll break.

maciuszek committed:
address todos

maciuszek committed:
fix comment

maciuszek committed:
ensure memory and flush management for timers

charlievieth commented on stats.go:
```diff
@@ -298,30 +301,95 @@ func (c *gauge) Value() uint64 {
+	if t.count < t.capacity {
+		t.values = append(t.values, value)
+	} else {
+		t.values = append(t.values[1:], value) // discard the oldest value when the reservoir is full, this can probably be smarter
```
Treating this more as a ring buffer when at capacity would be better since it would provide a more accurate representation of the most recent timings (whereas here we only update the first timing):

```diff
diff --git a/stats.go b/stats.go
index 70a582b..99a0a0a 100644
--- a/stats.go
+++ b/stats.go
@@ -346,6 +346,7 @@ type reservoirTimer struct {
 	capacity int
 	values   []float64
 	count    int
+	off      int
 	mu       sync.Mutex
 }
 
@@ -364,7 +365,8 @@ func (t *reservoirTimer) AddValue(value float64) {
 	if t.count < t.capacity {
 		t.values = append(t.values, value)
 	} else {
-		t.values = append(t.values[1:], value) // discard the oldest value when the reservoir is full, this can probably be smarter
+		t.values[t.off%len(t.values)] = value
+		t.off++
 	}
 
 	t.count++
```

charlievieth replied:
Also, making `capacity` a power of two would allow for us to use a mask instead of an expensive modulo operation. This can be cheaply calculated using the `bits` package:

```go
func nextPowerOfTwo(capacity int) uint {
	return 1 << bits.Len(uint(capacity))
}
```

charlievieth commented on stats.go:
```diff
@@ -298,30 +301,95 @@ func (c *gauge) Value() uint64 {
+	capacity int
+	values   []float64
+	count    int
+	mu       sync.Mutex
```
super nit, make the Mutex the first field (it's advantageous performance wise to make the most accessed field first)

charlievieth commented on stats.go:
```diff
@@ -336,6 +404,7 @@ func (ts *timespan) CompleteWithDuration(value time.Duration) {
 }
 
 type statStore struct {
+	// todo: no idea how memory is managed here, when are the map entries ever deleted?
```
The assumption is the that stat names are stable over the lifetime of a process, but this is a known issue for processes that incorrectly use this library and cause a cardinality explosion of stat names. See:

* https://github.com/lyft/gostats/pull/158 (original issue)
* https://github.com/lyft/gostats/pull/160 (my proposed fix)

charlievieth commented on net_sink.go:
```diff
@@ -260,15 +260,23 @@ func (s *netSink) FlushGauge(name string, value uint64) {
+}
+
+func (s *netSink) FlushAggregatedTimer(name string, value, sampleRate float64) {
+	suffix := fmt.Sprintf("|ms|@%.1f\n", sampleRate)
```
For performance reasons it would be great to avoid `fmt.Sprintf` here if possible.

charlievieth commented on stats.go:
```diff
@@ -393,6 +466,18 @@ func (s *statStore) Flush() {
+			for _, value := range timer.CollectedValue() {
+				s.sink.FlushAggregatedTimer(key.(string), value, sampleRate)
+			}
+			s.timers.Delete(key) // delete it from the map so it's not flushed again
```
Deletion is expensive and since we assume the names of stats are stable will just lead to it being re-added to the map at a later time. For counters we only flush non-zero values and I think we could do something similar with the `reservoirTimer` type (say not flush if it has zero values and trim the slice to zero elements on flush).

charlievieth commented on stats.go:
```diff
@@ -214,7 +214,10 @@ type StatGenerator interface {
-	return &statStore{sink: sink}
+	return &statStore{
+		sink: sink,
+		conf: GetSettings(), // todo: right now the environment is being loaded in multiple places and can be made more efficient
```
The [sync.OnceValue](https://pkg.go.dev/sync#OnceValue) function can help eliminate these multiple loads.

maciuszek committed:
optimize reservoirTimer by utilizing a ring buffer
optimize reservoirTimer by allocating an reusing memory
fix bug with sampleRate calculation
optimize FlushAggregatedTimer a bit
expand and fix test coverage

maciuszek commented:
@charlievieth just did a first pass of your comments. Sincerely thank you both for the clarification regarding how data/keys in the maps are reused, and the ring buffer recommendation.

I now have a lot of new todos I need to dive into, but I just pushed a pretty big change (ref: https://github.com/lyft/gostats/pull/171/commits/ea5ae6aacd66b6e97c948b2b2d7c2f7d833135a8), if/when you have a chance I would be eager to know if it aligns with your thoughts.

Fwiw I appreciate any and all nits. Very much a go noob.

Edit: had a logical bug which I addressed in https://github.com/lyft/gostats/pull/171/commits/e81d603dbad94e0eac4e1ed167a54a647735dce2

maciuszek committed:
correct how we flush reusable timer entries

maciuszek committed:
add test for reused timer map after flushing

charlievieth commented:
> @charlievieth just did a first pass of your comments. Sincerely thank you both for the clarification regarding how data/keys in the maps are reused, and the ring buffer recommendation.

np! I'm vacation the next few days but will take a look at your changes when I can.

charlievieth commented on stats.go:
```diff
@@ -298,30 +302,130 @@ func (c *gauge) Value() uint64 {
+	t.values[t.overflow&t.ringMask] = value
+	t.overflow++
+
+	// todo: can i optimize this with xor?
```
I don't think you need to reset the overflow since you're masking (taking only the lower bits is equivalent to a modulo operation). Also using an unsigned int here would be a bit safer / clearer (makes the value always positive and unsigned overflow is well defined).

maciuszek replied:
Ah this implementation here is wrong

maciuszek committed:
correct the ring buffer implementation to utilize bitwise benefits
additionally remove variablility of the reservoir size as we need to ensure this number is a power of two

maciuszek committed:
improve reservoirTimer property access

maciuszek committed:
make reservoir tests more dynamic
improve comments

maciuszek committed:
improve comments

maciuszek committed:
optimize reservoir timer flush

maciuszek committed:
block never flush edge cases when stores are constructed outside of NewDefaultStore and resevoir timers are enabled
implicitly optimize store construction

maciuszek committed:
fix typo in comment

maciuszek committed:
add test for reservoir automatic flushing
fix data race in reservoirTimer

sokada1221 commented on stats.go:
```diff
@@ -298,30 +310,125 @@ func (c *gauge) Value() uint64 {
+}
+
+func (t *reservoirTimer) Reset() {
+	atomic.StoreUint64(&t.count, 0)
```
Don't we need a lock here?

maciuszek replied:
It's already atomic so the access should be protected iiuc. i did add a test for data races https://github.com/lyft/gostats/pull/171/commits/976215245c1e4f9f51d5bd0bc4ff1691fc46dd46 , but actually this isn't covered, let me cover this reset too.

I had a data race that displayed how atomic works. Inside a mutex i didn't use atomic access, however there was access (atomic) to the same data from a different method outside of the mutex, the run down is that the access in the mutex also needed to be atomic and that itself protects from datarace

maciuszek replied:
I think this test should cover it https://github.com/lyft/gostats/pull/171/commits/264192494f7a8ae6f5c6d8c11d4c1c928fc81cd4 , this will aggressively flush (and Reset) with interleaving writes (changing the counter), to the same reservoirTimer allocation in the (statStore instance).timers

maciuszek replied:
Nevermind I think you're right! Thank you @sokada1221

maciuszek replied:
@sokada1221 https://github.com/lyft/gostats/pull/171/commits/4e9611db7de9ca240f59d30bb3c9cbc5d8ed3ac4

maciuszek committed:
add test for concurrent reservoir writes and flushing

maciuszek committed:
fix typo in comment

maciuszek committed:
protect writes while flushing

maciuszek committed:
dont export controls that can result in a deadlock or datarace

maciuszek committed:
add critical optimization todo

maciuszek committed:
simplify reservoir processing

maciuszek committed:
unexport RingSize and document immutability

maciuszek committed:
print to stdout for testing

maciuszek committed:
improve test logging

maciuszek committed:
temporarily make logging a bit better at the sacrifice of performance

maciuszek committed:
remove test logging
